### ISSUE TITLE
resizable bugs

### ISSUE THREAD

**Issue by damien-git** [deitch/jstree-grid#89]
In Chromium, when trying to resize a column, the column separator moves suddently to the left and then moves twice as fast as the mouse. The problem does not occur with Firefox for the same tree.
Also, resizable no longer works after the tree is destroyed and rebuilt, even with Firefox.

**Comment by deitch** [comment]
Resizing is such a pain. I think I spent more time on it than any other issue (second was column alignment, which is closely related).

I am seeing similar issues in Chrome.

**Comment by deitch** [comment]
No, my mistake, no issues in Chrome. 

What is the code you are using, and the Chromium version?

**Comment by damien-git** [comment]
I see bugs with the demo, using jsTree 3.0.9. But the demo seems to work with jsTree 3.0.8, so I guess there was a change in jsTree that triggered that bug. I'm using Chromium 41 on Kubuntu.
As a side note, I wish I did not have to give any width for the columns, and let the browsers adjust that like they would for a table: browsers are pretty good at it. With a good layout, users are less likely to want to resize columns.

**Comment by damien-git** [comment]
Actually, I am seeing the bug also when I use jsTree 3.0.8 in my application, so there is more. I will try to create a simple test case.

**Comment by damien-git** [comment]
I did some debugging. It looks like using toResize.prevCol.css("width") does not work with Chromium when a column size is set to 'auto' (it returns 0) (it works with Firefox). The result is that the total of the column widths that you set is a lot less than the table width. You might want to get a calculated width, for a cell if it doesn't work with col.

**Comment by deitch** [comment]
Let's see:

1. LL424-425 get the column to the left of the clicked separator (`prevCol`) and the one to the right (`nextCol`) on `mousedown`
2. LL397-398 determine the width of those columns on `mousemove`
3. LL401-403 calculate what the new widths should be
4. LL407-409 apply those new widths

You are saying that the old widths in LL397-398 are coming back as 0 because the column size is set to 'auto'? Is the problem on any particular column, or just the last? Quoted from the README.md L72:

> * width: width of the column in pixels. If no width is given, the default is 25 px **except for the last column**. In the last column, if not width is given, it is treated as 'auto' and fills the entire rest of the grid to the right.

Is the problem on any column or just the last? Maybe put up a quick jsfiddle so we are on the same page?

**Comment by damien-git** [comment]
> You are saying that the old widths in LL397-398 are coming back as 0 because the column size is set to 'auto'?

The ones for columns with a size set to 'auto'. In other words, if the previous column (toResize.prevCol) has a size set to 'auto', toResize.prevCol.css("width") will return 0.

> Is the problem on any particular column, or just the last?

It is the first column. I have not tried another one.

Here is a fiddle:  https://jsfiddle.net/umpgLL1x/

**Comment by deitch** [comment]
I see why it is a problem on Chrome. For some reason, whenever you set the `col` element to a certain width, it doubles the `th` width. I don't know if this is related to something here, or a Chrome thing in general. In FF, when I manually set the `col` to a particular width, the `th` and all of the `td` go with it appropriately. But in Chrome, it doubles it in some strange way. I am going to set up a fiddle just with this to see.

**Comment by deitch** [comment]
Wow, that is a difficult one to track down.

Here is the source of the problem; not quite sure how to fix it. 

The structure looks something like this (details left out):

````html
<div>
  <table>
    <colgroup>
       <col></col>
       <col></col>
    </colgroup>
   <tr><th></th><th></th></tr>
   <tr>
     <td>  <!-- tree goes here -->
     </td>
   </tr>
   <tr>
     <td>  <!-- grid cells go here -->
     </td>
   </tr>
  </table>
</div>
````

We set a width for the `col` elements as to how wide we want them to be.

* Firefox: we did not set a width for the `table` or the wrapping `div`, so they stretch (as do the `tr`) the entire available width. However, the actual elements - the `col` and `th` and `td` - respect the widths set on the columns, so all looks good.
* Chrome: we did not set a width for the `table` or the wrapping `div`, so they stretch (as do the `tr`) the entire available width. The `col` elements respect the widths set, but the `th` and `td` that match them do not so it all goes haywire.

In sum, FF and Chrome treat the columns differently if the sum of widths of `col` is different than the total table width.

FF behaves more as we expected, but Chrome actually has a correct internal logic.

The simplest thing would be to tell the table, "conform to the sum of all widths." But that would mess up auto.

Open to ideas....

**Comment by damien-git** [comment]
I do not understand:
- " whenever you set the col element to a certain width, it doubles the th width" (I did not notice that)
- how a col element could respect the width but not a corresponding td
- why we cannot just set the widths to avoid inconsistencies

What do you mean by "sum of widths" when one of them is 'auto' ?

Ok, one idea: when the user changes a cell width before or after a cell with a width set to 'auto', only change the width of the column that has a set width (so, either the one before or after). Or calculate the width of the 'auto' column based on the first td in the column (instead of the col).

**Comment by deitch** [comment]
@damien-git I guess I didn't make it clear. Forget the 'auto' problem for a second, it is a generalized issue with Chrome/Chromium and FF that 'auto' just highlights. Just focus on my most recent comment.

Look at the html in that comment. `div` wraps `table`, which wraps `colgroup` with `col` elements, and multiple `tr` with `td` or `th` elements.

Let's say the viewport is 500px wide. The `div` (unless otherwise explicitly instructed) will expand to 500px.

Inside the `div` is a `table`. The `table` (unless otherwise instructed) will automatically expand to fill the `div`, i.e. 500px (I am ignoring any padding and margins for now).

This means the `tr` elements will expand to 500px as well. 

Now you set the first `col` width to 100px and the second to 200px, e.g., so the total is 300px, but the entire table is 500px, so we have an unresolved 200px here.

What happens to the `td` and `th` elements in each row?

* Option A: ignore the `col` settings, just expand the `th` and `td` to fill up the `tr` width of 500px. This is what Chrome/Chromium does.
* Option B: respect the `col` settings, set the `th` and `td` to 100px and 200px respectively, and have the `tr` be much wider than the sum of `th` or `td`. This is what Firefox does.

All of the weirdness you are seeing is because we set the widths on the `col`, but Chrome/Chromium ignore those to fill up the space.

We need a way to tell Chrome, "shrink the table to fit the sum of the col widths, rather than expanding the cells to fill the tr which expands to fill the table, which expands to fill the div, which expands to fill the viewport."

When no column is set to 'auto', the problem is bad enough, but in theory resolvable, since we *know* what the total should be, as does Chrome. Just tell it to shrink (I don't know how, but is possible in theory). 

When a column is set to 'auto', it is a double-whammy. We cannot tell the browser to shrink a table that has a column told to expand to the table!

Some possiblilities:

* Always have at least one column as 'auto' (the last automatically if none provided)
* Require a width on the entire thing
* Other ideas?

**Comment by damien-git** [comment]
> The table (unless otherwise instructed) will automatically expand to fill the div

No. You have to set its width, otherwise browsers optimize the table width based on what is inside. Currently this is done with .jstree-grid-table setting the width to 100%.

> Now you set the first col width to 100px and the second to 200px, e.g., so the total is 300px, but the entire table is 500px, so we have an unresolved 200px here.

We should never be in that situation.

I think we should let developers decide the table width (not specified by default, as opposed to the current 100% by default) and the column widths (not specified by default). So that by default, browsers optimize the table layout in the best possible way. We let browsers decide what to do if there is an error in the configuration (such as a total width for columns superior to the table width).

Then, when a user starts to set a column width, we set all the column widths in pixels to their current values (as returned by getComputedStyle() or jquery's .css() on the first cell in the column), we unset the width for the table, and we adjust values for the column widths from there.

In short, we would interpret 'auto' as 'auto' only for the initial table layout.

**Comment by deitch** [comment]
>> The table (unless otherwise instructed) will automatically expand to fill the div

>No. You have to set its width, otherwise browsers optimize the table width based on what is inside. Currently this is done with .jstree-grid-table setting the width to 100%.

That is what I meant, but didn't explain well,:

> (because of width set to 100%) The table (unless otherwise instructed) will automatically expand to fill the div

You are suggesting the developer explicitly sets the table width. Where do they specify it? After all, there is no `table` in what they do; it is auto-generated. Options:

* parameter as part of jstree-grid configuration
* keep table at 100% of div which is 100% of wrapping element, but indicate in docs that it will fill whatever element is wrapping the jstree, so it is set outside

Personally, I like the second option, as it fits with the general html flow, and requires no code changes. In other words, the `table` will fill up your entire parent element width, so if you want to control it, set the width. That is pretty easy to do by adding a `div`. We just need to make sure the developer *knows* it, which is a change in docs; easy.

Which brings us back to the beginning. If you just set the wrapping element's width (or leave it, which expands to viewport), and set the widths of columns in a way that do not add up, and the browser handles it (e.g. `'auto'`), currently the calculations are messed up. We may need to fix just that.

But I am strongly disinclined to change the width on the table. I far prefer to leave it at 100% of the wrapping element, and let the developer set the width on the wrapping element.

**Comment by adamjimenez** [comment]
The issue I'm seeing is that the resize is out of step with the mouse. So one pixel of mouse movement resizes the column about 5 pixels. Also it would be cool if there was a persistent option to save the column widths to a cookie or localstorage.

**Comment by side-by-side** [comment]
I did fix this by adding two lines in the "mousemove" handler.
Maybe it's a dirty hack, but it works for me, at least for now.
https://github.com/deitch/jstree-grid/blob/master/jstreegrid.js#L408

```
if (diff !== 0){
    oldPrevHeaderInner = toResize.prevHeader.width();
    oldNextHeaderInner = toResize.nextHeader.width();
    oldPrevColWidth = parseFloat(toResize.prevCol.css("width"));
    oldNextColWidth = parseFloat(toResize.nextCol.css("width"));

    if (!oldPrevColWidth) oldPrevColWidth = toResize.prevHeader.innerWidth(); // new line
    if (!oldNextColWidth) oldNextColWidth = toResize.nextHeader.innerWidth(); // new line
```

In Chrome (as in Opera), if the column width is not defined or "auto", ```parseFloat(toResize.prevCol.css("width"))``` will return 0. In Firefox it returns actual width.
So. in Chrome we can get that width from the header.

**Comment by deitch** [comment]
@side-by-side cool. I am trying this right now.

**Comment by deitch** [comment]
The prize goes to @side-by-side! I am adding this in right now, along with a line in contributors. Thank you!

v3.2.7

**Comment by side-by-side** [comment]
And thank you for the very useful plugin for jsTree :)
